pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,346 @@
+		s, err := h.inner.NextSeries(ctx)
+		if err != nil {
+			if errors.Is(err, types.EOS) {
+				return fmt.Errorf("exhausted series before all groups were completed: %w", err)
+			}
+			return err
+		}
+		thisSeriesGroups := h.seriesGroups[h.currentInnerSeriesIndex]
+
+		// Native histograms only ever go to their original labelset.
+		// Floats only ever go to their group.
+		// It is possible that a series has both floats and histograms.
+		// It is also possible that both series groups are the same.
+		// The conflict in points is then detected in computeOutputSeriesForGroup.
+		h.saveNativeHistogramsToGroup(s.Histograms, thisSeriesGroups[0])
+		h.saveFloatsToGroup(s.Floats, thisSeriesGroups[1])
+		h.currentInnerSeriesIndex++
+	}
+	return nil
+}
+
+// saveFloatsToGroup places each fPoint into a bucket with the upperBound set by the input series.
+func (h *HistogramFunctionOverInstantVector) saveFloatsToGroup(fPoints []promql.FPoint, g *bucketGroup) {
+	if len(fPoints) > 0 {
+		upperBound, err := strconv.ParseFloat(h.bucketValues[h.currentInnerSeriesIndex], 64)
+		if err != nil {
+			// The le label was invalid. Record it:
+			h.Annotations.Add(annotations.NewBadBucketLabelWarning(
+				h.innerSeriesMetricNames.GetMetricNameForSeries(h.currentInnerSeriesIndex),
+				h.bucketValues[h.currentInnerSeriesIndex],
+				h.inner.ExpressionPosition(),
+			))
+		} else {
+			for _, f := range fPoints {
+				pointIdx := h.timeRange.PointIndex(f.T)
+				g.pointBuckets[pointIdx] = append(
+					g.pointBuckets[pointIdx],
+					bucket{
+						upperBound: upperBound,
+						count:      f.F,
+					},
+				)
+			}
+		}
+	}
+	// We are done with the fPoints, so return these now
+	types.FPointSlicePool.Put(fPoints, h.memoryConsumptionTracker)
+	g.remainingSeriesCount--
+}
+
+// saveNativeHistogramsToGroup stores the given native histograms onto the given group.
+// There should only ever be one native histogram per a series or series group. In general,
+// they are per-series, but we handle them as parts of groups because they could hypothetically
+// have the same series name as a classic histogram.
+func (h *HistogramFunctionOverInstantVector) saveNativeHistogramsToGroup(hPoints []promql.HPoint, g *bucketGroup) {
+	if len(hPoints) > 0 {
+		// We should only ever see one set of native histograms per group.
+		if g.nativeHistograms != nil {
+			panic("We should never see more than one native histogram per group")
+		}
+		g.nativeHistograms = &hPoints
+	}
+	// hPoint's are returned to the pool after computeOutputSeriesForGroup is finished with them
+	g.remainingSeriesCount--
+}
+
+func (h *HistogramFunctionOverInstantVector) computeOutputSeriesForGroup(_ context.Context, g *bucketGroup) (types.InstantVectorSeriesData, error) {
+	floatPoints, err := types.FPointSlicePool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	histogramIndex := 0
+	var currentHPoint *promql.HPoint
```

Comment:
[nit] This could move inside the `for` and could just be a `histogram.FloatHistogram`, I don't think we need the time stamp once we've identified the histogram is for this time step.

Reply:
We do because `histogramIndex` may move beyond the current step. However we can keep track of the `histogram.FloatHistogram` and `histogramIdx` separately which may make some of the other logic slightly nicer so I'll tidy that up.